gpt4all-api/gpt4all_api/app/api_v1/routes/chat.py

```diff
@@ -45,15 +47,41 @@ async def chat_completion(request: ChatCompletionRequest):
     '''
     Completes a GPT4All model response based on the last message in the chat.
     '''
-    # Example: Echo the last message content with some modification
+    # GPU is not implemented yet
+    if settings.inference_mode == "gpu":
+        raise HTTPException(status_code=400,
+              detail=f"Not implemented yet: Can only infere in CPU mode.")
+
+    # we only support the configured model
+    if request.model != settings.model:
+        raise HTTPException(status_code=400,
+              detail=f"The GPT4All inference server is booted to only infer: `{settings.model}`")
+
+    # run only of we have a message
     if request.messages:
-        last_message = request.messages[-1].content
-        response_content = f"Echo: {last_message}"
+        model = GPT4All(model_name=settings.model, model_path=settings.gpt4all_path)
+
+        # format system message and conversation history correctly
+        formatted_messages = ""
+        for message in request.messages:
+            formatted_messages += f"<|im_start|>{message.role}\n{message.content}<|im_end|>\n"
+
+        # the LLM will complete the response of the assistant
+        formatted_messages += "<|im_start|>assistant\n"
+        response = model.generate(
+            prompt=formatted_messages,
+            temp=request.temperature
+            )
+
+        # the LLM may continue to hallucinate the conversation, but we want only the first response
+        # so, cut off everything after first <|im_end|>
+        index = response.find("<|im_end|>")
+        response_content = response[:index].strip()
```

Comment:
Why is ChatML hard-coded here? Normally GPT4All models have a customizable prompt template that defaults to the value in [models2.json](https://gpt4all.io/models/models2.json).

Reply:
This is done tue to the very specifics of OpenAI's API. OpenAI's ChatCompletions endpoint can receive an array of messages containing the system message, and past user questions/statements and the assistants replies (see https://platform.openai.com/docs/api-reference/chat/create and https://cookbook.openai.com/examples/how_to_format_inputs_to_chatgpt_models):

```
curl https://api.openai.com/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "gpt-3.5-turbo",
    "messages": [
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": "Knock knock."},
            {"role": "assistant", "content": "Who's there?"},
            {"role": "user", "content": "Orange."},
    ]
  }'
```

Goal was here to parse the past messages and to provide the LLM with the complete history. From the above example, I am creating a result like this:

```
<|im_start|>system
You are a helpful assistant.<|im_end|>
<|im_start|>user
Knock knock.<|im_end|>
<|im_start|>assistant
Who's there?<|im_end|>
<|im_start|>user
Orange.<|im_end|>
<|im_start|>assistant
```

IMHO we would a different representation of prompt templates in model2.json to be able to reliably parse and use those for this specific use case... Until then, I decided to hardcode ChatML.